abdiff/core/calc_ab_diffs.py

```diff
@@ -61,23 +63,52 @@ def get_diffed_batches_iter(
 
         # convert batch to pandas dataframe and calc values for new columns
         df = batch.to_pandas()  # noqa: PD901
-        df["ab_diff"] = df.apply(
+
+        # calculate all diffs and unpack into separate columns
+        diff_results = df.apply(
             lambda row: calc_record_diff(row["record_a"], row["record_b"]), axis=1
         )
-        df["has_diff"] = df["ab_diff"].apply(lambda diff_value: diff_value != "{}")
+        df["ab_diff"] = diff_results.apply(lambda x: x[0])
+        df["modified_timdex_fields"] = diff_results.apply(
+            lambda x: list(x[1]) if x[1] else []
+        )
+        df["has_diff"] = diff_results.apply(lambda x: x[2])
 
         yield pa.RecordBatch.from_pandas(df)  # type: ignore[attr-defined]
 
 
-def calc_record_diff(record_a: bytes | None, record_b: bytes | None) -> str | None:
-    """Calculate symmetric diff from two JSON strings."""
+def calc_record_diff(
+    record_a: str | bytes | dict | None,
+    record_b: str | bytes | dict | None,
```

Comment:
Why the type hint change? It feel odd to account 3 different potential types for these `record_*` params, is there any way to normalize around one?

Reply:
Originally, the type signature was `bytes | None`.  We either got bytes when a record had the version (and it usually did) or handle situations where it was missing.  

Supporting `dict` as well made [testing easier](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/edc5f80dda71028adfc35bd2e658751054102634/tests/test_calc_ab_diffs.py#L18-L20) where we didn't have to constantly convert a dictionary to bytes, just to have this convert it back.

While we are expecting `bytes` from the input datasets, got to thinking there is no reason why it _can't_ also support `str` as well.  For a future developer, I think it should represent what it can accept without extra work or modification, and this includes `str` as well.

It's a good question, but in this instance, I don't think we benefit by limiting the allowed types to this function.  I think it's fairly common in conversion or serialization type libraries, where they can handle _lots_ of types of inputs generally.